action.yaml

```diff
@@ -141,12 +154,16 @@ runs:
     - name: Gather base file paths
       shell: bash
       id: filepaths
+      env:
+        scandir: ${{ inputs.scandir }}
+        exclude_args: ${{ steps.exclude.outputs.excludes }}
+        additional_file_args: ${{ steps.additional.outputs.files }}
```

Comment:
I think that this depends on the output. For example, in the case of `steps.exclude.outputs.excludes` and `steps.additional.outputs.files`, their values are a transformation of `inputs.ignore_*` and `inputs.additional_files` respectively.

If someone adds a malicious payload into one of those inputs, it can eventually executed when those outputs are injected.

Reply:
They do not?
They are now passed as env?

Reply:
I was referring to the hypothetical case where the code would use an intermediate env var for the input but not the output. I believe that's what you're suggesting here. My understanding is that you're saying that outputs don't need to pass through intermediate variables. I believe that they do need to.

Here's simplified snippet of what I mean. I'll take the `steps.options.outputs.options` example here.

```yml
    # ...
    - name: Set options
      shell: bash
      id: options
      env:
        # Note: Inputs passed through intermediate env vars
        severity: ${{ inputs.severity }}
        format: ${{ inputs.format }}
      run: |
        declare -a options
        if [[ -n "${severity}" ]]; then
          options+=("-S ${severity}")
        fi
        options+=("--format=${format}")
        echo "::set-output name=options::${options[@]}"

    # ...

    - name: Print information
      shell: bash
      # Note: Outputs passed as is with no intermediate variable as you're suggesting
      run: |
        echo "Files: ${{steps.filepaths.outputs.filepaths}}"
        echo "Excluded: ${{ steps.exclude.outputs.excludes }}"
        echo "Options: ${{ steps.options.outputs.options }}"
        echo "Status code: ${{steps.check.outputs.statuscode}}"
        exit ${{steps.check.outputs.statuscode}}
```

Getting code execution in this context can be done by calling the action like this:

```yaml
use: ludeeus/action-shellcheck@master
with:
  format: 'gcc";echo "injected code";echo "'
```

The injection would happen on the following line:

```sh
echo "Options: ${{ steps.options.outputs.options }}"
```

which would be interpreted as

```sh
echo "Options: --format=";echo "injected code"; echo ""
```

This example is theoretical as I haven't tested it. It may require an extra quote or backslash here and there.

I hope that illustrates my point that outputs that may contain user inputs need to be passed through intermediate env variables as well. For the protection against this kind of attack to work, intermediate variables need to be used for every value that may have included user input.